File: win/MarkdownWin/MarkdownWin/Account.cs
//
//  Account.cs
//  MarkdownWin
//
//  The signed-in person. Mirrors Account.swift.
//
//  There is no server behind this yet: signing in records a name and email locally so the
//  sidebar can show an account. Swapping in a real service means replacing LogIn and the
//  storage, not the views.
//

using System;
using System.Linq;

namespace MarkdownWin;

internal sealed class Account : ObservableBase
{
    private string fullName;
    private string email;

    public string FullName
    {
        get => fullName;
        private set => Set(ref fullName, value);
    }

    public string Email
    {
        get => email;
        private set => Set(ref email, value);
    }

    public bool IsLoggedIn => !string.IsNullOrEmpty(FullName);

    /// <summary>Initials for the sidebar's avatar, e.g. "Ada Lovelace" becomes "AL".</summary>
    public string Initials
    {
        get
        {
            string letters = string.Concat(FullName
                .Split(' ', StringSplitOptions.RemoveEmptyEntries)
                .Take(2)
                .Select(word => word[0]));
            return letters.Length == 0 ? "?" : letters.ToUpperInvariant();
        }
    }

    public Account()
    {
        fullName = AppSettings.AccountName ?? string.Empty;
        email = AppSettings.AccountEmail ?? string.Empty;
    }

    public void LogIn(string fullName, string email)
    {
        string name = fullName.Trim();
        if (name.Length == 0)
        {
            return;
        }

        FullName = name;
        Email = email.Trim();

        AppSettings.AccountName = FullName;
        AppSettings.AccountEmail = Email;

        Notify(nameof(IsLoggedIn));
        Notify(nameof(Initials));
    }

    public void LogOut()
    {
        FullName = string.Empty;
        Email = string.Empty;
        AppSettings.AccountName = null;
        AppSettings.AccountEmail = null;

        Notify(nameof(IsLoggedIn));
        Notify(nameof(Initials));
    }
}

File: macos/Markdown/Markdown/Account.swift
//
//  Account.swift
//  Markdown
//
//  The signed-in person.
//
//  There is no server behind this yet: signing in records a name and email locally so the
//  sidebar can show an account. Swapping in a real service means replacing `logIn` and the
//  storage, not the views.
//

import SwiftUI

@Observable
final class Account {
    private(set) var fullName: String
    private(set) var email: String

    var isLoggedIn: Bool { !fullName.isEmpty }

    /// Initials for the sidebar's avatar, e.g. "Ada Lovelace" becomes "AL".
    var initials: String {
        let letters = fullName
            .split(separator: " ")
            .prefix(2)
            .compactMap(\.first)
        return letters.isEmpty ? "?" : String(letters).uppercased()
    }

    init() {
        let defaults = UserDefaults.standard
        fullName = defaults.string(forKey: PreferenceKey.accountName) ?? ""
        email = defaults.string(forKey: PreferenceKey.accountEmail) ?? ""
    }

    func logIn(fullName: String, email: String) {
        let name = fullName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return }

        self.fullName = name
        self.email = email.trimmingCharacters(in: .whitespacesAndNewlines)

        UserDefaults.standard.set(self.fullName, forKey: PreferenceKey.accountName)
        UserDefaults.standard.set(self.email, forKey: PreferenceKey.accountEmail)
    }

    func logOut() {
        fullName = ""
        email = ""
        UserDefaults.standard.removeObject(forKey: PreferenceKey.accountName)
        UserDefaults.standard.removeObject(forKey: PreferenceKey.accountEmail)
    }
}
